clippy_lints/src/methods/unnecessary_iter_cloned.rs

```diff
@@ -40,6 +40,56 @@ pub fn check_for_loop_iter(
         && !clone_or_copy_needed
         && let Some(receiver_snippet) = snippet_opt(cx, receiver.span)
     {
+        // Issue 12098
+        // https://github.com/rust-lang/rust-clippy/issues/12098
+        // if the assignee have `mut borrow` conflict with the iteratee
+        // the lint should not execute, former didn't consider the mut case
+
+        // check whether `expr` is mutable
+        fn is_mutable(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+            if let Some(hir_id) = path_to_local(expr)
+                && let Node::Pat(pat) = cx.tcx.hir_node(hir_id)
+            {
+                matches!(pat.kind, PatKind::Binding(BindingAnnotation::MUT, ..))
+            } else {
+                true
+            }
+        }
+
+        fn is_caller_or_fields_changed(cx: &LateContext<'_>, body: &Expr<'_>, caller: &Expr<'_>) -> bool {
+            if let ExprKind::Block(block, ..) = body.kind {
+                for stmt in block.stmts {
+                    if let StmtKind::Semi(e) = stmt.kind {
+                        // dont check `is_mutable` for `lhs` here because
+                        // assign to an immutable variable is an error
+                        if let Some(assignee) = match e.kind {
+                            ExprKind::Assign(assignee, _, _) | ExprKind::AssignOp(_, assignee, _) => Some(assignee),
+                            _ => None,
+                        } {
+                            if !can_mut_borrow_both(cx, caller, assignee) {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+            false
+        }
+
+        if let ExprKind::Call(_, [child, ..]) = expr.kind {
+            // filter first layer of iterator
+            let mut child = child;
+            // get inner real caller requests for clone
+            while let ExprKind::MethodCall(_, caller, _, _) = child.kind {
+                child = caller;
+            }
+            if is_mutable(cx, child) && is_caller_or_fields_changed(cx, body, child) {
+                // skip lint
+                return true;
+            }
+        };
```

Comment:
I'm surprised `rustfmt` isn't complaining about this colon. Oh well, if it's not complaining... :)

Reply:
The semicolon might be needed, if the if statement has a type other than `()`. `rustfmt` doesn't use type information for formatting, AFAIK. Therefore, it cannot safely remove this one.

We could build a Clippy lint for this. We already have the counter part (and one of my favorite lints): [clippy::semicolon_if_nothing_returned](https://rust-lang.github.io/rust-clippy/master/index.html#/semicolon_if_nothing_returned)